prov/gni/test/allocator.c

```diff
@@ -0,0 +1,203 @@
+	if (!fd) {
+		fprintf(stderr, "error opening /proc/self/maps.\n");
+		return ret;
+	}
+
+	line = malloc(size);
+	if (!line) {
+		fprintf(stderr, "error mallocing space for line.\n");
+		return ret;
+	}
+
+	while (getline(&line, &size, fd) != -1) {
+		if (strstr(line, "hugetlbfs")) {
+			ret = 1;
+			break;
+		}
+	}
+
+	free(line);
+	fclose(fd);
+
+	return ret;
+}
+
+TestSuite(mbox_creation, .init = allocator_setup, .fini = allocator_teardown);
+
+Test(mbox_creation, alloc_single_page)
+{
+	int ret;
+
+	ret = gnix_mbox_allocator_create(ep_priv->nic, NULL, GNIX_PAGE_4MB,
+					 100, 100, &allocator);
+	assert_eq(ret, FI_SUCCESS, "gnix_mbox_allocator_create failed.");
+	expect_eq(verify_hugepages(), 1,
+		  "memory not found in /proc/self/maps.");
+
+	ret = gnix_mbox_allocator_destroy(allocator);
+	assert_eq(ret, FI_SUCCESS, "gnix_mbox_allocator_destroy failed.");
+	expect_eq(verify_hugepages(), 0,
+		  "memory not released in /proc/self/maps.");
+}
+
+Test(mbox_creation, alloc_three_pages)
+{
+	int ret;
+
+	ret = gnix_mbox_allocator_create(ep_priv->nic, NULL, GNIX_PAGE_4MB,
+					 1000, 12000, &allocator);
+	assert_eq(ret, FI_SUCCESS, "gnix_mbox_allocator_create failed.");
+	expect_eq(verify_hugepages(), 1,
+		  "memory not found in /proc/self/maps.");
+
+	ret = gnix_mbox_allocator_destroy(allocator);
+	assert_eq(ret, FI_SUCCESS, "gnix_mbox_allocator_destroy failed.");
+	expect_eq(verify_hugepages(), 0,
+		  "memory not released in /proc/self/maps.");
+}
+
+Test(mbox_creation, alloc_mbox)
+{
+	int ret;
+
+	struct gnix_mbox *mail_box;
+	struct slist_entry *entry;
+	struct gnix_slab *slab;
+
+	char test_string[] = "hello allocator.";
+
+	ret = gnix_mbox_allocator_create(ep_priv->nic, NULL, GNIX_PAGE_4MB,
+					 1000, 12000, &allocator);
+	assert_eq(ret, FI_SUCCESS, "gnix_mbox_allocator_create failed.");
+	expect_eq(verify_hugepages(), 1,
+		  "memory not found in /proc/self/maps.");
+
+	ret = gnix_mbox_alloc(allocator, &mail_box);
+	expect_eq(ret, FI_SUCCESS, "gnix_mbox_alloc failed.");
+
+	expect(mail_box);
+
+	entry = allocator->slab_list.head;
+	assert(entry);
+
+	slab = container_of(entry, struct gnix_slab, list_entry);
+
+	expect_eq(mail_box->slab, slab,
+		  "slab list head and mail box slab pointer are not equal.");
+	expect_eq(mail_box->memory_handle, &mail_box->slab->memory_handle,
+		 "mail_box memory handle not equal to slab memory handle.");
+	expect_eq(mail_box->offset, 0, "offset is not 0.");
+	expect_eq(mail_box->base, mail_box->slab->base,
+		  "mail_box base not equal to slab base.");
+
+	memcpy(mail_box->base, test_string, sizeof(test_string));
+	expect_strings_eq((char *) mail_box->base, test_string);
+
+	ret = gnix_mbox_free(mail_box);
+	expect_eq(ret, FI_SUCCESS, "gnix_mbox_free failed.");
+
+	ret = gnix_mbox_allocator_destroy(allocator);
+	assert_eq(ret, FI_SUCCESS, "gnix_mbox_allocator_destroy failed.");
+	expect_eq(verify_hugepages(), 0,
+		  "memory not released in /proc/self/maps.");
+}
```

Comment:
Could you add more tests to ensure the error conditions go off as intended for the testable cases? EINVAL,EIO, etc. 

Memory allocation failure testing is a nightmare. I don't recommend writing tests for that.